packages/vite/src/node/server/index.ts

```diff
@@ -503,20 +503,25 @@ export async function createServer(
   middlewares.use(serveRawFsMiddleware(server))
   middlewares.use(serveStaticMiddleware(root, server))
 
-  // spa fallback
-  if (!middlewareMode || middlewareMode === 'html') {
-    middlewares.use(spaFallbackMiddleware(root))
-  }
-
-  // run post config hooks
-  // This is applied before the html middleware so that user middleware can
-  // serve custom content instead of index.html.
+  // We need to apply post server hooks before `spaFallbackMiddleware()`.
```

Comment:
I tried to make this change in an earlier PR but @patak-dev reported that it broke Vitepress: https://github.com/vitejs/vite/pull/4670#issuecomment-907748589. Perhaps it's okay to change in a major version though? @patak-dev would probably know better as I'm not familiar with Vitepress

Reply:
Made a Vitepress PR to fix the situation: https://github.com/vuejs/vitepress/pull/625.